packages/jupyter-ai/jupyter_ai/context_providers/base.py

```diff
@@ -0,0 +1,216 @@
+        self._chat_history = chat_history
+        self.llm_chat_memory = llm_chat_memory
+        self.root_dir = os.path.abspath(os.path.expanduser(root_dir))
+        self.preferred_dir = get_preferred_dir(self.root_dir, preferred_dir)
+        self.dask_client_future = dask_client_future
+        self.chat_handlers = chat_handlers
+        self.context_providers = context_providers
+
+        self.llm = None
+
+    @abc.abstractmethod
+    async def make_context_prompt(self, message: HumanChatMessage) -> str:
+        """Returns a context prompt for all commands of the context provider
+        command.
+        """
+        pass
+
+    def replace_prompt(self, prompt: str) -> str:
+        """Modifies the prompt before sending it to the LLM."""
+        return prompt
+
+    def _clean_prompt(self, text: str) -> str:
+        # util for cleaning up the prompt before sending it to a retriever
+        for provider in self.context_providers.values():
+            text = provider.replace_prompt(text)
+        return text
+
+    @property
+    def base_dir(self) -> str:
+        # same as BaseChatHandler.output_dir
+        if self.preferred_dir and os.path.exists(self.preferred_dir):
+            return self.preferred_dir
+        else:
+            return self.root_dir
+
+    def get_llm(self):
+        lm_provider = self.config_manager.lm_provider
+        lm_provider_params = self.config_manager.lm_provider_params
+
+        curr_lm_id = (
+            f'{self.llm.id}:{lm_provider_params["model_id"]}' if self.llm else None
+        )
+        next_lm_id = (
+            f'{lm_provider.id}:{lm_provider_params["model_id"]}'
+            if lm_provider
+            else None
+        )
+
+        if not lm_provider or not lm_provider_params:
+            return None
+
+        if curr_lm_id != next_lm_id:
+            model_parameters = self.model_parameters.get(
+                f"{lm_provider.id}:{lm_provider_params['model_id']}", {}
+            )
+            unified_parameters = {
+                "verbose": True,
+                **lm_provider_params,
+                **model_parameters,
+            }
+            llm = lm_provider(**unified_parameters)
+            self.llm = llm
+        return self.llm
+
+
+class ContextCommand(BaseModel):
+    cmd: str
+
+    @property
+    def id(self) -> str:
+        return self.cmd.partition(":")[0]
+
+    @property
+    def arg(self) -> Optional[str]:
+        if ":" not in self.cmd:
+            return None
+        return self.cmd.partition(":")[2].strip("'\"").replace("\\ ", " ")
+
+    def __str__(self) -> str:
+        return self.cmd
+
+    def __hash__(self) -> int:
+        return hash(self.cmd)
+
+
+class BaseCommandContextProvider(BaseContextProvider):
+    id_prefix: ClassVar[str] = "@"
+    only_start: ClassVar[bool] = False
+    requires_arg: ClassVar[bool] = False
+    remove_from_prompt: ClassVar[bool] = (
+        False  # whether the command should be removed from prompt
+    )
```

Comment:
Docstrings should also go here; the docstrings should explain what each attribute does and why it exists. Including examples is also OK. Feel free to make them up to a paragraph long if you need to.